src/mon/commands/Command.h

```diff
@@ -0,0 +1,333 @@
+
+
+/* Command
+ *
+ * Define a command. It will be templated with
+ *  T	    = a paxos service
+ *  Stable  = the typename of the paxos service's stable map
+ *  Pending = the typename of the paxos service's pending map
+ *
+ * Depending on a paxos service's specificities, Stable and Pending may be
+ * the same.
+ *
+ * This class doesn't use these types for much more than defining the prepare
+ * and preprocess functions to be called by derived classes, but still
+ * requires them to be defined to properly abstract command definitions.
+ */
+
+typedef std::shared_ptr<Formatter> FormatterRef;
+
+template<typename T, typename Stable, typename Pending>
+struct Command : public Common
+{
+
+  explicit Command(
+      Monitor *_mon,
+      PaxosService *_service,
+      CephContext *_cct) :
+    Common(_mon, _service, _cct)
+  { }
+  
+  virtual ~Command() { };
+
+  virtual bool preprocess(
+      MonOpRequestRef op,
+      const cmdmap_t &cmdmap,
+      const Stable &stable_map) {
+
+    stringstream ss;
+    bufferlist rdata;
+
+    string format;
+    cmd_getval(cct, cmdmap, "format", format, string("plain"));
+    FormatterRef f(Formatter::create(format));
+
+    string prefix;
+    cmd_getval(cct, cmdmap, "prefix", prefix);
+
+    return do_preprocess(op, prefix, cmdmap, ss, rdata, f, stable_map);
+  }
+
+  virtual bool prepare(
+      MonOpRequestRef op,
+      const cmdmap_t &cmdmap,
+      Pending &pending_map,
+      Stable &stable_map) {
+
+    op->mark_osdmon_event(__func__);
+    stringstream ss;
+    bufferlist rdata;
+
+    string format;
+    cmd_getval(cct, cmdmap, "format", format, string("plain"));
+    FormatterRef f(Formatter::create(format));
+
+    string prefix;
+    cmd_getval(cct, cmdmap, "prefix", prefix);
+
+    return do_prepare(op, prefix, cmdmap,
+	ss, rdata, f, pending_map, stable_map);
+  }
+
+  virtual bool do_preprocess(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      const Stable &stable_map) = 0;
+
+  virtual bool do_prepare(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      Pending &pending_map,
+      Stable &stable_map) = 0;
+
+  virtual bool handles_command(const string &prefix) = 0;
+
+ private:
+  Command() = delete;
+};
+
+
+/* ReadCommand
+ *
+ * Defines a templated Read-only command. These will be called during
+ * preprocess. This class abstracts a derived class' command definition, by
+ * redefining the preprocess command from the 'Command' class, and then
+ * calling a function to be implemented by the class deriving from this one.
+ *
+ * Preparing is still left defined as the 'Command' class has it defined,
+ * which is basically returning 'false'.
+ */
+template<typename T, typename Stable, typename Pending>
+struct ReadCommand : public Command<T, Stable, Pending>
```

Comment:
can we mark `WriteCommand ` `final`? as it is not supposed to be inherited, i guess?

Reply:
@jecluis have you considered migrating some of the read commands into mgr to implement them in python?

Reply:
I haven't, no. Maybe that's something we could explore after this is merged? Don't know what kind of effort would be required, and unless there's a strong argument to do that at the same time as this work, I'd rather do that later.